src/lib/Imp.hs

```diff
@@ -275,7 +274,7 @@ lookupVar v = do
     Just v' -> v'
 
 indexSetSize :: Type -> ImpM IExpr
-indexSetSize (IdxSetLit n) = return $ ILit (IntLit n)
+indexSetSize (FixedIntRange low high) = return $ ILit $ IntLit $ high - low
```

Comment:
What's interesting is that this case seems unnecessary, but in fact programs as simple as `[1, 2, 3]` fail to compile without it! This is because if it's skipped, then each time we query this function (e.g. to generate the shape of an array), we allocate a fresh variable and use that in the shape expression. This causes some type mismatches later, because even though those variables originate from the same expression, we use different variables and our type-level equality is purely syntactic. To better illustrate this, lowering a program like `[1, 2, 3]` without this extra rule produces this:
```
> Compiler bug!
> Please report this at github.com/google-research/dex-lang/issues
>
> Type mismatch in copy: (Int, [v_2]) != (Int, [v])(, copy v_1 v_3)
> === context ===
> Checking after imp pass:
> imports: ()
> [v_1:Ptr (Int[v])]
> v:Int = %isub 3 0
> v_2:Int = %isub 3 0
> v_3:Ptr (Int[v_2]) = alloc (Int, [v_2])
> v_4:Ptr (Int[]) = v_3.0
> store v_4 1
> v_5:Ptr (Int[]) = v_3.1
> store v_5 2
> v_6:Ptr (Int[]) = v_3.2
> store v_6 3
> copy v_1 v_3
> free v_3
> (*ans* @> L (afor *:3 . aget v_1))
> exports: (*ans* @> L (3=>Int))
```

Note that what I did in here is not a robust fix, and the same problem will bite us in the future when we will be interested in compiling truly dependent ranges, because then we have no choice other than doing the subtraction at runtime. One way to fix this that I thought about would be to memoize answers of `indexSetSize` inside `ImpM` so that each type will get assigned a unique size variable that will be used throughout the lowered program. WDYT?

Reply:
Turns out that another way to solve this problem is to add a simplification case to `impSub`. I'll do that in a followup PR that I'm preparing now.

Reply:
That sounds like a good fix, at least until we think about it more. Basic constant folding seems like a very reasonable thing to do short of full canonicalization.